[v5.8] Add migration code for BoltDB to SQLite

mheon opened the pull request:
This is gated behind a new option in `podman system migrate`, `--migrate-db`.

The basic logic is simple:
* Podman is already configured to use BoltDB
* Open a new, fresh SQLite database to write into
* Migrate all database contents as they exist in BoltDB, to SQLite. ** Do this as simply as possible: grab the object from the old DB
   and write it into the new DB using the standard Add and Save
   functions.
* Set the new database in the Runtime, close the old one.
* Move the old database file so it won't be reused
* Show a warning if the user explicitly configured BoltDB in containers.conf

Our ability to test complex migration scenarios is limited, but this should handle simple migrations easily.

<!--
Thanks for sending a pull request!

For more detailed information, please review our contributing guidelines:
https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests
-->

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:
<!-- Use [x] to mark as done, or click the checkbox after opening PR -->

- [ ] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [ ] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [ ] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [ ] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [ ] All commits pass `make validatepr` (format/lint checks)
- [ ] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

<!--
Write `None` if there are no user-facing changes, otherwise enter your release note below.
Include "action required" if users need to take action when upgrading.
-->

```release-note
Added the ability to migrate from the legacy BoltDB database to SQLite via a new option to `podman system migrate`, `--migrate-db`.
```

mheon committed:
Add migration code for BoltDB to SQLite

This is gated behind a new option in `podman system migrate`,
`--migrate-db`.

The basic logic is simple:
* Podman is already configured to use BoltDB
* Open a new, fresh SQLite database to write into
* Migrate all database contents as they exist in BoltDB, to
  SQLite.
** Do this as simply as possible: grab the object from the old DB
   and write it into the new DB using the standard Add and Save
   functions.
* Set the new database in the Runtime, close the old one.
* Move the old database file so it won't be reused
* Show a warning if the user explicitly configured BoltDB in
  containers.conf

Our ability to test complex migration scenarios is limited, but
this should handle simple migrations easily.

Fixes #27628

Signed-off-by: Matt Heon <matthew.heon@pm.me>

Luap99 commented on libpod/runtime_migrate.go:
```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+
+		// Does the SQLite state already exist?
+		dbPath := filepath.Join(sqliteStateBasePath(r), "db.sql")
+		if _, err := os.Stat(dbPath); err == nil {
```
we use fileutils.Exists everywhere for such checks

Luap99 commented on libpod/runtime_migrate.go:
```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+		if err := os.Rename(boltDBPath, newBoltDBPath); err != nil {
+			return fmt.Errorf("renaming old database %s to %s: %w", boltDBPath, newBoltDBPath, err)
+		}
+		logrus.Errorf("Old database has been renamed to %s and will no longer be used", newBoltDBPath)
```
Why is this an error? There was no failure here so seeing error will confuse people.

Either info level (which we don't show by default) or to show it by default I could live with a println.

I don't think there is much value to show this by default if the migration worked. The man page can mention how to recover (or delete) the old db file.

Luap99 commented on libpod/runtime_migrate.go:
```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+		}
+		if backend == config.DBBackendBoltDB {
+			logrus.Errorf("In containers.conf, database_backend is manually set to \"boltdb\" - please change to \"sqlite\" to complete migration")
+		}
```
This is wrong as it will always trigger

`runtime.config.Engine.DBBackend = config.DBBackendBoltDB.String()` is set explicitly  in getDBState() when the config file has the default.

I believe you would need to use config.Default(), though even then I am not sure since it might have overwritten the cached default. Maybe a config.New() to actually read the config file values from disk again.

Luap99 reviewed (commented):
The alive lock is taken after we have the db state of course. Which means in practise the alive lock isn't of help at all. Because what this would mean is all commands blocked on the alive lock will then use the wrong boltdb state directly after they got the lock post migrate.
However we cannot take the alive lock earlier as it depends on the db path validation on the db.

I wonder if there needs to be some additional fault check when we got the alive lock that knows the migration just happen and then it either manges do redo the state in sqlite or just exit with error and let the command be retired.

So the lock would only reduce the risk of parallel writes but of course even that it cannot do for long running podman processes

mheon commented on libpod/runtime_migrate.go:
```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+		}
+
+		// Does the SQLite state already exist?
+		dbPath := filepath.Join(sqliteStateBasePath(r), "db.sql")
```
I think returning two strings, base path and filename, might be best, as the actual SQLite code gets a bit convoluted with how it builds the final path,

mheon commented on libpod/runtime_migrate.go:
```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+		}
+		if ctrError != nil {
+			return ctrError
+		}
```
Were you thinking of 1 function for the whole migration, or a function each for migrating containers/migrating pods/migrating volumes?

mheon commented:
Would make testing manually a pain, though. Only thing I can think of against it.

Unless we retain the option for manual migration, in case folks don't want to restart?

debarshiray reviewed (commented):
I am interested in how the BoltDB to SQLite migration code is meant to be used from programs that wrap Podman, like Toolbx.

>> do we actually need the `--db` part? it would seem we already know the db and can detect and migrate it automatically. I see the precedent but just curious.
>
> I'd prefer to require an explicit argument so we can error if they try and migrate with an existing SQLite DB

My initial thought was to agree with @baude about not needing the `--migrate-db` option because of the way Toolbx has been invoking `podman system migrate` for a while now.  It's possible that I am getting this wrong, so here are some details.

Every time the `toolbox(1)` binary is invoked, it uses `podman version ...` to compare the current Podman version with the last seen version, stored in a stamp file, and runs `podman system migrate` if the current one is newer and updates the last seen version.

This means that Toolbx needs to conditionally use the `--migrate-db` option based on some checks.  Will `podman info --format '{{.Host.DatabaseBackend}}'` be the right check for that?  Does running `podman system migrate --migrate-db` include everything that happens without the option or do we need to run both?

mheon commented on docs/source/markdown/podman-system-migrate.1.md:
```diff
@@ -26,6 +26,12 @@ newly configured mappings.
+
+Migrate from the legacy BoltDB database to SQLite.
+Support for BoltDB will be removed in Podman 6.0.
+Podman will display a warning if this migration is necessary.
```
Done

Luap99 commented:
> I am interested in how the BoltDB to SQLite migration code is meant to be used from programs that wrap Podman, like Toolbx.
> 
> > > do we actually need the `--db` part? it would seem we already know the db and can detect and migrate it automatically. I see the precedent but just curious.
> > 
> > 
> > I'd prefer to require an explicit argument so we can error if they try and migrate with an existing SQLite DB
> 
> My initial thought was to agree with @baude about not needing the `--migrate-db` option because of the way Toolbx has been invoking `podman system migrate` for a while now. It's possible that I am getting this wrong, so here are some details.
> 
> Every time the `toolbox(1)` binary is invoked, it uses `podman version ...` to compare the current Podman version with the last seen version, stored in a stamp file, and runs `podman system migrate` if the current one is newer and updates the last seen version.
> 

I think we talked about this, you really should not be doing this. migrate is meant to be run without other podman commands in parallel generally (you cannot guarantee that), it stops all containers (toolbx shouldn't do that either), and it hasn't been needed since some very early 1.X days.


If we go down the auto refresh thing on boot nobody would need to do anything and we can just have the migrate option for manual testing.

mheon commented:
You know, this really should not be working. We're syncing Pod state before creating containers, but pod state has the infra container's ID in it - which, in SQLite, is a foreign key on container ID. So SQLite should be refusing to commit our pod state update because the infra container ID refers to a container ID that doesn't exist (normally, the infra container ID only gets set after the infra container is created). Doesn't directly impact this PR but needs to be looked into.

mheon committed:
Deterministically order pod inspect fields

There are two fields I'm worried about: shared namespaces and pod
containers. Both are generated via loops over maps and are thus
non-deterministic in ordering. Throw a sort on each to fix the
order so we can actually diff `podman pod inspect` output.

Signed-off-by: Matt Heon <mheon@redhat.com>

Luap99 commented on test/system/555-migration.bats:
```diff
@@ -0,0 +1,86 @@
+    # Perform migration
+    if [[ "$CI_DESIRED_DATABASE" == "boltdb" ]]; then
+        # Podman will print an extra warning here because containers.conf is set to boltdb
+        run_podman 0+w $safe_opts system migrate --migrate-db
```
ah this seems to print an error not an warning which needs +e instead of +w I belive

Luap99 reviewed (approved):
LGTM, but we really need people to test this code extensively ideally before we push that into all the "stable" repos

lsm5 reviewed (commented):
LGTM
